src/helpers/http/mod.rs

```diff
@@ -0,0 +1,202 @@
+mod network;
+
+pub use network::HttpNetwork;
+
+use crate::{
+    ff::Field,
+    helpers::{messaging::Gateway, Direction, Error, GatewayConfig, Role},
+    net::{
+        discovery::{peer, PeerDiscovery},
+        BindTarget, MessageSendMap, MpcHelperServer,
+    },
+    protocol::{
+        boolean::random_bits_generator::RandomBitsGenerator, context::SemiHonestContext, prss,
+        QueryId,
+    },
+    task::JoinHandle,
+};
+use rand::thread_rng;
+use std::net::SocketAddr;
+
+pub struct HttpHelper {
+    role: Role,
+    peers: [peer::Config; 3],
+    gateway_config: GatewayConfig,
+    participant: prss::Endpoint,
+    server: MpcHelperServer,
+}
+
+impl HttpHelper {
+    pub fn new<D: PeerDiscovery>(
+        role: Role,
+        peer_discovery: &D,
+        gateway_config: GatewayConfig,
+    ) -> Self {
+        // Prss
+        let peers = peer_discovery.peers();
+        let participant_setup = prss::Endpoint::prepare(&mut thread_rng());
+        let participant = participant_setup.setup(
+            &peers[role.peer(Direction::Left)].prss.public_key,
+            &peers[role.peer(Direction::Right)].prss.public_key,
+        );
+
+        Self {
+            role,
+            peers,
+            gateway_config,
+            participant,
+            server: MpcHelperServer::new(MessageSendMap::default()),
+        }
+    }
+
+    /// binds server as defined in the peers config list
+    /// # Panics
+    /// if peers config does not specify port
+    pub async fn bind(&self) -> (SocketAddr, JoinHandle<()>) {
+        let this_conf = &self.peers[self.role];
+        let port = this_conf.http.origin.port().unwrap();
+        let target = BindTarget::Http(format!("127.0.0.1:{}", port.as_str()).parse().unwrap());
+        tracing::info!("starting server; binding to port {}", port.as_str());
+        self.server.bind(target).await
+    }
+
+    /// adds a query to the running server so that it knows where to send arriving data
+    /// # Errors
+    /// if a query has been previously added
+    pub fn query(&self, query_id: QueryId) -> Result<Gateway, Error> {
```

Comment:
I looked at this interface again and I am not sure I understand why it needs to be like this. When you have a new query, the thing I would want to start the protocol is context, not gateway. why we need to keep two separate methods?

Reply:
main thing is the context takes a reference to a gateway, not its value. where would that value be kept if I returned context directly from this `query` fn?

Reply:
today we discussed the need for clients to initiate new query by running prss protocol first. to do that they would need to borrow the gateway, so we should keep it somewhere. 

Can you briefly summarize here the responsibilities of each component that you're building here? I see MpcHelperServer and HttpHelper but not sure how is everything is going to work together as an application. Who is responsible of keeping track of running queries and executing prss negotiation to create context?

Reply:
Here's what my idea is:

we are creating a new API that a "report collector" (or test runner) will call to initiate a new query. That request handling will create the query, initiate prss, and run the protocol. everything would be kept as local variables in this context.

since we don't want to keep the request open after we empty the body, we could just spawn a new task that continues on after the body is emptied. Or maybe just send the body on a channel back to the `HttpHelper` that has an event loop that spawns the task. That might be easier since the event loop could have easy access to the functions on `HttpHelper`.

One piece that i'm not sure about is how do we handle multiple steps of the protocol? If we want to run those steps simultaneously, you'd need a way to run them independently, so maybe spawn a task for each step?

If any failure happens in the middle, then you would need a way to stop all spawned tasks. it would be possible with an event loop per task that watches for a close signal, and the failing task sends the signal to all others.

An alternative is to add a map to the `HttpHelper`. something like `query_id -> runtime_components`. so all of the spawned tasks' `JoinHandle`s are stored in the `runtime_components`. Now, if one of the tasks fails, there could be a mechanism to simply `abort` all of the handles.

In either case, the piece responsible for initiating the query would effectively be the request handler for the "create query" request.